### PR TITLE
[dev-v2.8] Add K8s 1.27.5

### PR DESCRIPTION
**Pull request by rayandas**
Related issue: https://github.com/rancher/rancher/issues/41840

### CODE REVIEW

File: pkg/images/images.go [hunk 1]
```diff
@@ -57,7 +57,7 @@ func GenerateRegSyncFile() {
 		}
 		images, err := getImages(distro, versions)
 		if err != nil {
-			logrus.Fatalf("failed to get images for %s: %v", distro, err)
+			//logrus.Fatalf("failed to get images for %s: %v", distro, err)
```
Comment (jiaqiluo) [inline pkg/images/images.go]: Question: Why are this and a few others skipped?

File: pkg/rke/k8s_rke_system_images.go [hunk 2]
```diff
@@ -9342,6 +9342,51 @@ func loadK8sRKESystemImages() map[string]v3.RKESystemImages {
 			WindowsPodInfraContainer:  "rancher/mirrored-pause:3.7",
 			Nodelocal:                 "rancher/mirrored-k8s-dns-node-cache:1.22.20",
 		},
+
+		// Enabled in Rancher v2.7.x
+		"v1.27.2-rancher2-1": {
+			Etcd:                      "quay.io/coreos/etcd:v3.5.7",
```
Comment (kinarashah) [inline pkg/rke/k8s_rke_system_images.go]: we need to use the mirrored image here

File: pkg/rke/k8s_rke_system_images.go [hunk 3]
```diff
@@ -9342,6 +9342,51 @@ func loadK8sRKESystemImages() map[string]v3.RKESystemImages {
...
+			NginxProxy:                "rancher/rke-tools:v0.1.89",
+			CertDownloader:            "rancher/rke-tools:v0.1.89",
+			KubernetesServicesSidecar: "rancher/rke-tools:v0.1.89",
+			KubeDNS:                   "rancher/mirrored-k8s-dns-kube-dns:1.22.20",
+			DNSmasq:                   "rancher/mirrored-k8s-dns-dnsmasq-nanny:1.22.20",
+			KubeDNSSidecar:            "rancher/mirrored-k8s-dns-sidecar:1.22.20",
+			KubeDNSAutoscaler:         "rancher/mirrored-cluster-proportional-autoscaler:1.8.6",
+			Flannel:                   "rancher/mirrored-flannel-flannel:v0.21.4",
+			FlannelCNI:                "rancher/flannel-cni:v0.3.0-rancher8",
+			CalicoNode:                "rancher/mirrored-calico-node:v3.25.0",
```
Comment (kinarashah) [inline pkg/rke/k8s_rke_system_images.go]: calico has a new version out, v3.26.1 so we'd need to update with introducing new minor v1.27

File: pkg/rke/k8s_rke_system_images.go [hunk 4]
```diff
@@ -9342,6 +9342,51 @@ func loadK8sRKESystemImages() map[string]v3.RKESystemImages {
...
+			WeaveNode:                 "weaveworks/weave-kube:2.8.1",
+			WeaveCNI:                  "weaveworks/weave-npc:2.8.1",
+			AciCniDeployContainer:     "noiro/cnideploy:5.2.3.6.1d150da",
+			AciHostContainer:          "noiro/aci-containers-host:5.2.3.6.1d150da",
+			AciOpflexContainer:        "noiro/opflex:5.2.3.6.1d150da",
+			AciMcastContainer:         "noiro/opflex:5.2.3.6.1d150da",
+			AciOpenvSwitchContainer:   "noiro/openvswitch:5.2.3.6.1d150da",
+			AciControllerContainer:    "noiro/aci-containers-controller:5.2.3.6.1d150da",
+			AciGbpServerContainer:     "noiro/gbp-server:5.2.3.6.1d150da",
+			AciOpflexServerContainer:  "noiro/opflex-server:5.2.3.6.1d150da",
```
Comment (jiaqiluo) [inline pkg/rke/k8s_rke_system_images.go]: not sure if **5.2.3**.6.1d150da1 is a newer tag than **5.2.7**.1.81c2369 used in the template for `v1.26.7-rancher1-1`.

File: pkg/rke/k8s_rke_system_images.go [hunk 5]
```diff
@@ -9342,6 +9342,51 @@ func loadK8sRKESystemImages() map[string]v3.RKESystemImages {
...
+			PodInfraContainer:         "registry.k8s.io/pause:3.9",
+			Ingress:                   "rancher/nginx-ingress-controller:nginx-1.7.0-rancher1",
```
Comment (kinarashah) [inline pkg/rke/k8s_rke_system_images.go]: we'd need a new PR under rancher/ingress-nginx for v1.8.x

**Review by kinarashah — COMMENTED** [review]
@rayandas along with mentioned addon version bumps, could you also double check other addon versions and see if they need to be updated?

File: pkg/rke/k8s_service_options.go [hunk 6]
```diff
@@ -526,12 +534,33 @@ func getKubeControllerOptions() map[string]string {
 	}
 }
 
+func getKubeControllerOptionsFor127() map[string]string {
```
Comment (jiaqiluo) [inline pkg/rke/k8s_service_options.go]: Comparing this map with the one returned by the existing `getKubeControllerOptions` method, it drops the pair `"pod-eviction-timeout": "5m0s",`. Why not reuse the existing method and handle the dropping in `getKubeControllerOptions127`?

File: pkg/rke/k8s_version_info.go [hunk 7]
```diff
@@ -656,6 +656,10 @@ func loadK8sVersionInfo() map[string]v3.K8sVersionInfo {
 			MinRancherVersion: "2.7.5-patch0",
 			MinRKEVersion:     "1.4.0-rc0",
 		},
+		"v1.27": {
+			MinRancherVersion: "2.7.5-patch0",
```
Comment (jiaqiluo) [inline pkg/rke/k8s_version_info.go]: Need confirmation for whether k8s 1.27 should be available in Rancher v2.7.5 or v2.7.6

File: regsync.yaml [hunk 8]
```diff
@@ -10,23 +10,6 @@ defaults:
     - application/vnd.oci.image.manifest.v1+json
     - application/vnd.oci.image.index.v1+json
 sync:
-  - source: docker.io/rancher/calico-cni
```
Comment (jiaqiluo) [inline regsync.yaml]: We do not expect any deletions on this file. Why does it happen?

**Review by jiaqiluo — CHANGES_REQUESTED** [review]

File: pkg/rke/k8s_rke_system_images.go (hunk 2 above)
Comment (rayandas) [reply pkg/rke/k8s_rke_system_images.go]: Updated.

File: pkg/images/images.go (hunk 1 above)
Comment (rayandas) [reply pkg/images/images.go]: I used my registry hence I had to comment this line. But now I am using rancher registry so reverted this comment.

File: regsync.yaml (hunk 8 above)
Comment (rayandas) [reply regsync.yaml]: Updated.

File: pkg/rke/k8s_rke_system_images.go (hunk 5 above)
Comment (rayandas) [reply pkg/rke/k8s_rke_system_images.go]: Raised the ingress nginx PR

File: pkg/rke/k8s_rke_system_images.go (hunk 3 above)
Comment (rayandas) [reply pkg/rke/k8s_rke_system_images.go]: Working on calico canal templates. will raise a PR soon

File: pkg/rke/k8s_rke_system_images.go (hunk 4 above)
Comment (rayandas) [reply pkg/rke/k8s_rke_system_images.go]: Updated the versions

File: pkg/rke/k8s_service_options.go (hunk 6 above)
Comment (rayandas) [reply pkg/rke/k8s_service_options.go:537]: I am looking into it.

Comment (rayandas) [reply pkg/rke/k8s_service_options.go]: Updated the PR with your suggestion. Could you please check? @jiaqiluo

File: pkg/rke/k8s_version_info.go (hunk 7 above)
Comment (kinarashah) [reply pkg/rke/k8s_version_info.go]: k8s 1.27 issue is for 2.8 milestone, so this would be v2.8.0 and RKE v1.5.x...

Comment (rayandas) [reply pkg/rke/k8s_version_info.go]: Alright! Thanks. Will update the PR

File: channels-rke2.yaml [hunk 9]
```diff
@@ -14,6 +14,8 @@ appDefaults:
         defaultVersion: '1.25.x'
       - appVersion: '>= 2.7.5-0 < 2.7.100-0'
         defaultVersion: '1.26.x'
+      - appVersion: '>= 2.7.5-0 < 2.7.100-0'
```
Comment (kinarashah) [inline channels-rke2.yaml]: I think we'd need to revise min/max channel server version and this range depending on target milestone

Comment (rayandas) [reply channels-rke2.yaml]: Update it to `- appVersion: '>= 2.8.0-0 < 2.8.100-0'` for now

**Comment by rayandas** [comment]
I have removed RKE2 and K3s changes from this PR because I was having some RKE2 chart related issues in CI (hence the CI was failing continuously). It would be good if RKE2 and K3s team takes care of 1.27.x support in KDM.
cc: @kinarashah

**Review by jiaqiluo — DISMISSED** [review]
LGTM.

**Comment by kinarashah** [comment]
@rayandas could you create a separate GH issue to introduce rke2/k3s v1.27.x version in KDM? This PR can be approved and merged with just RKE1 changes.

**Review by jiaqiluo — DISMISSED** [review]

File: pkg/rke/k8s_rke_system_images.go [hunk 10]
```diff
@@ -9342,6 +9342,51 @@ func loadK8sRKESystemImages() map[string]v3.RKESystemImages {
 			WindowsPodInfraContainer:  "rancher/mirrored-pause:3.7",
 			Nodelocal:                 "rancher/mirrored-k8s-dns-node-cache:1.22.20",
 		},
+
+		// Enabled in Rancher v2.8.x
+		"v1.27.5-rancher1-1": {
+			Etcd:                      "rancher/mirrored-coreos-etcd:v3.5.7",
+			Kubernetes:                "rancher/hyperkube:v1.27.5-rancher1",
+			Alpine:                    "rancher/rke-tools:v0.1.89",
```
Comment (kinarashah) [inline pkg/rke/k8s_rke_system_images.go]: We need to bump rke-tools to v0.1.90 for https://github.com/rancher/rancher/issues/42103 L9350-L9353

Comment (rayandas) [reply pkg/rke/k8s_rke_system_images.go]: Ah! just a sec. Updating

Comment (rayandas) [reply pkg/rke/k8s_rke_system_images.go]: @kinarashah Updated the PR

**Review by kinarashah — APPROVED** [review]

**Pull request merged**